### ISSUE TITLE
Do you have any stand-alone drm/msm source/patch?

### ISSUE THREAD

**Issue by Icenowy** [freedreno-zz/kernel-msm#2]
I want to apply freedreno to a current kernel provided by the vendor (already fixed and hacked on https://github.com/Icenowy/linux-kernel-u8800-aosc-oses), but I don't think I'm able to port this repo (kernel-msm) to the device, but port the driver back to the kernel.

**Comment by robclark** [comment]
from a quick look at that branch, it looks like a 2.6.35 based kernel.

I have backported the upstream driver as far back as 3.4 (which I haven't done for a while since there is already considerable difference between 3.4 (pre-devicetree) and upstream (devicetree, common clock framework, etc)..

You could look, for example, at 3.4 based ifc6410-drm branch.. that has drm backport plus a bunch of other dependencies backported.  But I guess it would be quite some work since 2.6.35 is even older and you'd have to work out what other dependent patches need to be backported..

**Comment by robclark** [comment]
hmm, I guess I should also note that the upstream driver does not have any adreno a2xx support currently.  Mostly for lack of any device with a2xx that could run a modern kernel and time.

**Comment by Icenowy** [comment]
Thanks for you help. I'll soon bump to an APQ8060 device( yep,this Galaxy SII HD). I may use hybris to drive U8800.

But should I do a 'git diff' between your kernel tree with and without drm? It seems that the upstream kernel of SHV-E120s provided by Samsung is also too old...

from Galaxy SII HD

-------- 原始邮件 --------
发件人： Rob Clark notifications@github.com 
日期: 2015/06/17  6:24 上午  (GMT+08:00) 
收件人： freedreno/kernel-msm kernel-msm@noreply.github.com 
抄送： Icenowy Zheng icenowy@outlook.com 
主题： Re: [kernel-msm] Do you have any stand-alone drm/msm source/patch? (#2) 

hmm, I guess I should also note that the upstream driver does not have any adreno a2xx support currently. Mostly for lack of any device with a2xx that could run a modern kernel and time.

**Comment by Icenowy** [comment]
@robclark I recently got a 3.4 for U8800. (https://github.com/Blefish/android_kernel_huawei_msm7x30) How can I port freedreno on it?

**Comment by robclark** [comment]
well, if you have a 3.4 based kernel, you could probably cherry-pick patches from the 3.4 based ifc6410-drm branch.. there would still be the matter of adding support for a2xx (which should be another subclass of adreno_gpu in the way that a3xx_gpu and a4xx_gpu are).  Note however that a2xx is using an mmu inside the gpu block (rather than an external iommu) so support for that would also need to be added.  There is the msm_mmu base class in drm/msm and the intention was that the a2xx mmu would be another subclass of that..

I assume that 7x30 is using the mdp4 display controller block.. although you'd have to add appropriate config bits in mdp4_get_config().

There is a branch somewhere w/ DSI support for mdp4 although 7x30 probably has a different dsi phy compared to 8x64..

**Comment by Icenowy** [comment]
Yes, it uses mdp4. Can I get enough infomation from the stock drivers/gpu/msm and drivers/video/msm? Applying the patch is very easy, far beyond my expectation! However, I found memset.S and dma-attrs.h already patched in the msm7x30 kernel by Blefish.Patching on the board-xxx.c failed, but Huawei devices uses a different board-xxx.c to other devices. (board-hw7x30*.c), so it means nothingHowever, I remembered that freedreno is first declared to support the adreno220 GPU shipped in msm8x60. (The current mobile phone which I use every day is a Galaxy SII HD, which uses APQ8060, so I noticed that)

Date: Thu, 18 Jun 2015 16:35:02 -0700

**Comment by robclark** [comment]
yeah, you should be able to figure out enough from the downstream android fbdev/gpu drivers..  I did find that the mdp4 code especially was a bit, umm, spaghetti code, so I found it easier to figure out how it worked by tracing the register reads/writes, etc.. see https://github.com/freedreno/kernel-msm/commits/ifc6410-vidre for the hacks I used.  In https://github.com/freedreno/envytools/ there is a demsm tool to decode the registers into bitfields which will possibly be useful (note: iirc you might have to edit the log to s/MDP/MDP4/ or something approximately like that for demsm to work.. at some point when I started working on mdp5 I renamed MDP to MDP4 in the xml files which have the register definitions)..

the gpu code in drivers/gpu/msm wasn't quite as messy as the drivers/video/msm code, so it should hopefully be easier to understand (but I still did similar trick for logging readl/writel)..

fwiw, the first freedreno gpu support was a2xx, but at that time I was using the kgsl (drivers/gpu/msm) kernel code.. I wanted to add a2xx support in drivers/gpu/drm/msm but never found time (and lack of modern kernel supporting a2xx devices plus huge amount of work needed for a3xx didn't help for finding free time to work on drm/msm a2xx support)

**Comment by Icenowy** [comment]
https://github.com/msm7x30/android_kernel_qcom_msm7x30Someone is working for a 3.10 kernel on msm7x30 Maybe Huawei devices are recommended, as it's maintained by LegacyHuawei's authors.

Date: Thu, 18 Jun 2015 18:33:07 -0700
